### PR TITLE
TidbMonitor add remotewrite configuration

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve?
TidbMonitor add remotewrite configuration.

### What is changed and how does it work?
Add remotewrite field into tidbmonitor spec.
### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test
- [ ] E2E test 
- [ ] Manual test 
### Side effects

Nonne

### Release Notes

```release-note
TidbMonitor support  `remotewrite` configuration
```

### CODE REVIEW

**Comment by mikechengwei** [comment]
Manual test screenshot：
image

File: pkg/monitor/monitor/monitor_manager.go [hunk 1]
```diff
@@ -275,6 +275,7 @@ func (m *MonitorManager) syncTidbMonitorConfig(tc *v1alpha1.TidbCluster, dc *v1a
 	}
 
 	newCM, err := getMonitorConfigMap(tc, dc, monitor)
+	klog.Errorf("tm monitor remote : %v", monitor.Spec.RemoteWrite[0])
```
Comment (zjj2wry) [inline pkg/monitor/monitor/monitor_manager.go:278]: Why output log here？

File: pkg/monitor/monitor/util.go [hunk 2]
```diff
@@ -139,6 +142,58 @@ func getMonitorConfigMap(tc *v1alpha1.TidbCluster, dc *v1alpha1.DMCluster, monit
 		EnableTLSDMCluster: dc != nil && dc.IsTLSClusterEnabled(),
 	}
 
+	if monitor.Spec.RemoteWrite != nil && len(monitor.Spec.RemoteWrite) > 0 {
+		var remoteWriteConfigs []*config.RemoteWriteConfig
+		for _, remoteWrite := range monitor.Spec.RemoteWrite {
+			url, err := client.ParseHostURL(remoteWrite.URL)
+			if err != nil {
+				continue
```
Comment (zjj2wry) [inline pkg/monitor/monitor/util.go:150]: would be better to output a warning log?

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 3]
```diff
@@ -291,3 +297,92 @@ type SecretOrConfigMap struct {
...
+	// +optional
+	BearerTokenFile string `json:"bearerTokenFile,omitempty"`
+	// TLS Config to use for remote write.
+	// +optional
+	TLSConfig TLSConfig `json:"tlsConfig,omitempty"`
+	// Proxy url
+	// +optional
+	ProxyURL string `json:"proxyUrl,omitempty"`
+	// +optional
+	QueueConfig QueueConfig `json:"queue_config,omitempty"`
```
Comment (zjj2wry) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:328]: should be a pointers type?

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 4]
```diff
@@ -291,3 +297,92 @@ type SecretOrConfigMap struct {
...
+	// File to read bearer token for remote write.
+	// +optional
+	BearerToken string `json:"bearerToken,omitempty"`
+	// +optional
+	// File to read bearer token for remote write.
+	// +optional
+	BearerTokenFile string `json:"bearerTokenFile,omitempty"`
+	// TLS Config to use for remote write.
+	// +optional
+	TLSConfig TLSConfig `json:"tlsConfig,omitempty"`
```
Comment (zjj2wry) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:323]: should be a pointers type?

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 5]
```diff
@@ -291,3 +297,92 @@ type SecretOrConfigMap struct {
...
+	// The URL of the endpoint to send samples to.
+	URL string `json:"url"`
+	// +optional
+	RemoteTimeout model.Duration `json:"remote_timeout,omitempty"`
+	// The list of remote write relabel configurations.
+	// +optional
+	WriteRelabelConfigs []RelabelConfig `json:"writeRelabelConfigs,omitempty"`
+	//BasicAuth for the URL.
+	// +optional
+	BasicAuth BasicAuth `json:"basicAuth,omitempty"`
```
Comment (zjj2wry) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:313]: should be a pointers type?

File: pkg/monitor/monitor/util.go [hunk 6]
```diff
@@ -139,6 +142,58 @@ func getMonitorConfigMap(tc *v1alpha1.TidbCluster, dc *v1alpha1.DMCluster, monit
...
+					Action:       writeRelabelConfig.Action,
+				})
+			}
+			remoteWriteConfigs = append(remoteWriteConfigs, &config.RemoteWriteConfig{
+				URL:                 &config.URL{URL: url},
+				RemoteTimeout:       remoteWrite.RemoteTimeout,
+				WriteRelabelConfigs: writeRelabelConfigs,
+				HTTPClientConfig:    httpClientConfig,
+				QueueConfig: config.QueueConfig{
+					Capacity:          remoteWrite.QueueConfig.Capacity,
```
Comment (zjj2wry) [inline pkg/monitor/monitor/util.go:184]: If not specified queue config, Capacity、MaxShards... will set to 0?

**Comment by zjj2wry** [comment]
@mikechengwei would be better add a unittest for it

File: pkg/monitor/monitor/util.go (hunk 6 above)
Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:184]: image

**Comment by mikechengwei** [comment]
> @mikechengwei would be better add a unittest for it

ok

File: pkg/monitor/monitor/util.go (hunk 2 above)
Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:150]: util file can’t import log class.

File: pkg/monitor/monitor/util.go (hunk 6 above)
Comment (zjj2wry) [reply pkg/monitor/monitor/util.go:188]: We should only assign `Capacity、MaxShards...` value when QueueConfig is not nil，currently `remoteWrite.QueueConfig.Capacity` values are 0?

Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:188]: image

Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:188]: I think if we not set  `remoteWrite.QueueConfig.Capacity` field , it will use default value.

**Comment by DanielZhangQD** [comment]
@zjj2wry @csuzhangxc PTAL.

File: examples/monitor/tidb-monitor.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    baseImage: grafana/grafana
+    version: 6.1.6
+  initializer:
+    baseImage: registry.cn-beijing.aliyuncs.com/tidb/tidb-monitor-initializer
+    version: v4.0.9
+  reloader:
+    baseImage: registry.cn-beijing.aliyuncs.com/tidb/tidb-monitor-reloader
+    version: v1.0.1
+  imagePullPolicy: IfNotPresent
+  remoteWrite:
```
Comment (zjj2wry) [inline examples/monitor/tidb-monitor.yaml:21]: this field should be moving into `prometheus: {}`? WDYT? @csuzhangxc @mikechengwei

**Comment by zjj2wry** [comment]
Leave a question, generate LGTM

File: examples/monitor/tidb-monitor.yaml (hunk 7 above)
Comment (mikechengwei) [reply examples/monitor/tidb-monitor.yaml]: done.

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 8]
```diff
@@ -291,3 +297,92 @@ type SecretOrConfigMap struct {
...
+	// File to read bearer token for remote write.
+	// +optional
+	BearerToken string `json:"bearerToken,omitempty"`
+	// +optional
+	// File to read bearer token for remote write.
+	// +optional
+	BearerTokenFile string `json:"bearerTokenFile,omitempty"`
+	// TLS Config to use for remote write.
+	// +optional
+	TLSConfig *TLSConfig `json:"tlsConfig,omitempty"`
```
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:323]: Also the tls config in L195.

[the remark this answers is not in the document]
Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:323]: Yes. We need to support secret mount or configmap mount

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 9]
```diff
@@ -291,3 +297,92 @@ type SecretOrConfigMap struct {
...
 	ConfigMap *corev1.ConfigMapKeySelector `json:"configMap,omitempty"`
 }
+
+// RemoteWriteSpec defines the remote_write configuration for prometheus.
+// +k8s:openapi-gen=true
+type RemoteWriteSpec struct {
+	// The URL of the endpoint to send samples to.
+	URL string `json:"url"`
+	// +optional
+	RemoteTimeout model.Duration `json:"remote_timeout,omitempty"`
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:307]: How about unifying the tag format to `remoteTimeout`?

Comment (csuzhangxc) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:307]: ditto for `queue_config`.

Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:307]: ok

File: pkg/monitor/monitor/util.go [hunk 10]
```diff
@@ -139,6 +142,10 @@ func getMonitorConfigMap(tc *v1alpha1.TidbCluster, dc *v1alpha1.DMCluster, monit
 		EnableTLSDMCluster: dc != nil && dc.IsTLSClusterEnabled(),
 	}
 
+	if monitor.Spec.Prometheus.RemoteWrite != nil && len(monitor.Spec.Prometheus.RemoteWrite) > 0 {
```
Comment (csuzhangxc) [inline pkg/monitor/monitor/util.go:145]: only use `len(monitor.Spec.Prometheus.RemoteWrite) > 0` is engouh?

File: pkg/monitor/monitor/util.go [hunk 11]
```diff
@@ -1170,3 +1177,90 @@ func buildExternalLabels(monitor *v1alpha1.TidbMonitor) model.LabelSet {
 	}
 	return m
 }
+
+func generateRemoteWrite(monitor *v1alpha1.TidbMonitor) []*config.RemoteWriteConfig {
+	var remoteWriteConfigs []*config.RemoteWriteConfig
+	for _, remoteWrite := range monitor.Spec.Prometheus.RemoteWrite {
+		url, err := client.ParseHostURL(remoteWrite.URL)
+		if err != nil {
+			continue
```
Comment (csuzhangxc) [inline pkg/monitor/monitor/util.go:1186]: log any error? ditto for L1215?

Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:1186]: Util class file can‘t log.

Comment (csuzhangxc) [reply pkg/monitor/monitor/util.go:1186]: I found a `klog.Errorf` in this file, 😢

Comment (mikechengwei) [reply pkg/monitor/monitor/util.go:1186]: Ok. I will add.

File: pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go [hunk 12]
```diff
@@ -291,3 +297,92 @@ type SecretOrConfigMap struct {
...
+}
+
+// RelabelConfig allows dynamic rewriting of the label set, being applied to samples before ingestion.
+// It defines `<metric_relabel_configs>`-section of Prometheus configuration.
+// More info: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#metric_relabel_configs
+// +k8s:openapi-gen=true
+type RelabelConfig struct {
+	// A list of labels from which values are taken and concatenated
+	// with the configured separator in order.
+	SourceLabels model.LabelNames `json:"source_labels,omitempty"`
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go:350]: do we need to rename these tags?

Comment (mikechengwei) [reply pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go]: done.

File: examples/monitor-prom-remotewrite/tidb-monitor.yaml [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  prometheus:
+    baseImage: prom/prometheus
+    version: v2.18.1
+    remoteWrite:
+      - url: "http://localhost:1234"
+        writeRelabelConfigs:
+          - sourceLabels: [__test]
+            separator: ;
+            regex: (.*)
+            targetlabel: node
```
Comment (csuzhangxc) [inline examples/monitor-prom-remotewrite/tidb-monitor.yaml:17]:
```suggestion
            targetLabel: node
```

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**